abci/app.go

```diff
@@ -0,0 +1,378 @@
+				if err != nil {
+					panic(err)
+				}
+				respTxs[i] = &types.ExecTxResult{Code: CodeTypeOK}
+				app.state.Size++
+				// app.stagedBanTxs = append(app.stagedBanTxs, tx)
+			}
+
+		} else {
+			// There should never be a ban transaction after other txs
+			processedBanTxs = true
+			msg, err := model.ParseMessage(tx)
+			if err != nil {
+				respTxs[i] = &types.ExecTxResult{Code: CodeTypeEncodingError}
+			} else {
+				// Check if this sender already existed; if not, add the user too
+				err := UpdateOrSetUser(app.state.DB, msg.Sender, false, app.onGoingBlock)
+				if err != nil {
+					panic(err)
+				}
+				// Add the message for this sender
+				message, err := model.AppendToExistingMsgs(app.state.DB, *msg)
+				if err != nil {
+					return nil, err
+				}
+				app.onGoingBlock.Set([]byte(msg.Sender+"msg"), []byte(message))
+				chatHistory, err := model.AppendToChat(app.state.DB, *msg)
+				if err != nil {
+					panic(err)
+				}
+				// Append messages to chat history
+				app.onGoingBlock.Set([]byte("history"), []byte(chatHistory))
+				// This adds the user to the DB, but the data is not committed nor persisted until Comit is called
+				respTxs[i] = &types.ExecTxResult{Code: abci.CodeTypeOK}
+				app.state.Size++
+			}
+		}
+
+	}
+	app.state.Height = req.Height
+
+	response := &abci.ResponseFinalizeBlock{TxResults: respTxs, AppHash: app.state.Hash()}
+	return response, nil
+}
+
+// Commit the state and return the application Merkle root hash
+// Here we actually write the staged transactions into the database.
+// For details on why it has to be done here, check the Crash recovery section
+// of the ABCI spec
+func (app ForumApp) Commit(_ context.Context, commit *abci.RequestCommit) (*abci.ResponseCommit, error) {
+	if err := app.onGoingBlock.Commit(); err != nil {
+		panic(err)
+	}
+	saveState(&app.state)
+	return &abci.ResponseCommit{}, nil
+}
+
+// State Sync Connection
+// List available snapshots
+func (ForumApp) ListSnapshots(_ context.Context, listsnapshot *abci.RequestListSnapshots) (*abci.ResponseListSnapshots, error) {
+	return &abci.ResponseListSnapshots{}, nil
+}
+
+func (ForumApp) OfferSnapshot(_ context.Context, offersnapshot *abci.RequestOfferSnapshot) (*abci.ResponseOfferSnapshot, error) {
+	return &abci.ResponseOfferSnapshot{}, nil
+}
+
+func (ForumApp) LoadSnapshotChunk(_ context.Context, loadsnapshotchunk *abci.RequestLoadSnapshotChunk) (*abci.ResponseLoadSnapshotChunk, error) {
+	return &abci.ResponseLoadSnapshotChunk{}, nil
+}
+
+func (ForumApp) ApplySnapshotChunk(_ context.Context, applysnapshotchunk *abci.RequestApplySnapshotChunk) (*abci.ResponseApplySnapshotChunk, error) {
+	return &abci.ResponseApplySnapshotChunk{}, nil
+}
+
+func (app ForumApp) ExtendVote(_ context.Context, extendvote *abci.RequestExtendVote) (*abci.ResponseExtendVote, error) {
+	fmt.Println("Entered extend vote")
+
+	return &abci.ResponseExtendVote{VoteExtension: []byte(app.CurseWords)}, nil
```

Comment:
We should check somewhere (InitChain?) that we're not fed a config file with repeated curse words

Reply:
Yes this is one check I did not get a chance to include.